app/code/Magento/ImportExport/Model/Import/Source/FileParser/ZipParserFactory.php

```diff
@@ -0,0 +1,118 @@
+ */
+
+namespace Magento\ImportExport\Model\Import\Source\FileParser;
+
+use Magento\Framework\App\Filesystem\DirectoryList;
+use Magento\Framework\Filesystem;
+
+class ZipParserFactory implements ParserFactoryInterface
+{
+    private $filesystem;
+    private $isZipAvailable;
+    private $parserFactory;
+
+    public function __construct(
+        Filesystem $filesystem,
+        ParserFactoryInterface $parserFactory,
+        $isZipAvailable = null
+    ) {
+        $this->filesystem = $filesystem;
+        $this->parserFactory = $parserFactory;
+        $this->isZipAvailable = $isZipAvailable ?? extension_loaded('zip');
+    }
+
+    public function create($path, array $options = [])
+    {
+        $this->assertZipFileIsParsable($path);
+
+        $sourceDirectory = $this->filesystem->getDirectoryRead(DirectoryList::ROOT);
+        $writeDirectory = $this->filesystem->getDirectoryWrite(DirectoryList::TMP);
+
+        $path = $sourceDirectory->getRelativePath($path);
+        $this->assertFileExistance($path, $sourceDirectory);
+
+        $zip = new \ZipArchive();
+
+        if ($zip->open($sourceDirectory->getAbsolutePath($path)) !== true) {
+            throw new CorruptedFileException($path);
+        }
+
+        $files = $this->fetchFileNamesFromZipFile($zip);
+
+        foreach ($files as $file) {
+            $destinationFilename = $this->extractFileIntoDirectory($zip, $writeDirectory, $file);
+
+            try {
+                $parser = $this->parserFactory->create(
+                    $destinationFilename,
+                    ['directory_code' => DirectoryList::TMP] + $options
+                );
+            } catch (UnsupportedPathException $e) {
+                continue;
+            } finally {
+                $writeDirectory->delete($destinationFilename);
+            }
+
+            return $parser;
+        }
+
+        throw new UnsupportedPathException($path);
+    }
+
+    private function assertZipFileIsParsable($path)
+    {
+        if (substr($path, -4) !== '.zip') {
+            throw new UnsupportedPathException($path);
+        }
+
+        if (!$this->isZipAvailable) {
+            throw new UnsupportedPathException($path, 'Zip extension is not available');
+        }
+    }
+
+    private function assertFileExistance($path, $directory)
+    {
+        if (!$directory->isFile($path)) {
+            throw new UnsupportedPathException($path);
+        }
+    }
+
+    private function fetchFileNamesFromZipFile(\ZipArchive $zip)
+    {
+        $files = [];
+
+        for ($fileIndex = 0; $fileIndex < $zip->numFiles; $fileIndex++) {
+            $fileName = $zip->getNameIndex($fileIndex);
+            if ($this->isDirectoryZipEntry($fileName) || !$fileName) {
+                continue;
+            }
+
+            $files[] = $fileName;
+        }
+
+        return $files;
+    }
+
+    private function isDirectoryZipEntry($fileName)
+    {
+        return substr($fileName, -1) === '/';
```

Comment:
Did you see result of `pathinfo` with proper ZIP file list information?
It does not work, here is a proof inlined:
```
<?php

var_dump(pathinfo('path/to/a/zip/dir/'));

array(3) {
  ["dirname"]=>
  string(13) "path/to/a/zip"
  ["basename"]=>
  string(3) "dir"
  ["filename"]=>
  string(3) "dir"
}
```
As you see path info treats dir in the same way as a file. This makes impossible to use it for this functionality. If I use it breaks the tests as it is unaware of ZIP file format specifics.

`DIRECTORY_SEPARATOR` is a bad idea as well, as it is OS specific, but ZIP file format is not. Functionality will be broken if run on Windows.